protocol-messages.go

```diff
@@ -0,0 +1,380 @@
+package mls
+
+import (
+	"fmt"
+	"github.com/bifurcation/mint/syntax"
+)
+
+////// ClientInitKey
+
+// enum {
+//	invalid(0),
+//	supported_versions(1),
+//	supported_ciphersuites(2),
+//	expiration(3),
+//	(65535)
+// } ExtensionType;
+
+// struct {
+//	ExtensionType extension_type;
+//	opaque extension_data<0..2^16-1>;
+// } Extension;
+
+type ExtensionType uint16
+type Extension struct {
+	ExtensionType ExtensionType
+	ExtensionData []byte `tls:"head=2"`
+}
+
+// struct {
+//	ProtocolVersion supported_version;
+//	opaque client_init_key_id<0..255>;
+//	CipherSuite cipher_suite;
+//	HPKEPublicKey init_key;
+//	Credential credential;
+//	Extension extensions<0..2^16-1>;
+//	opaque signature<0..2^16-1>;
+// } ClientInitKey;
+type ClientInitKey struct {
+	ProtocolVersion uint8
+	CipherSuite     CipherSuite
+	HPKEPublicKey   []byte `tls:"head=2"`
+	Credential      Credential
+	Extensions      []Extension `tls:"head=2"`
+	Signature       []byte      `tls:"head=2"`
+}
+
+//// Proposal
+type ProposalType uint8
+
+const (
+	invalid = iota
+	add
+	update
+	remove
+)
+
+type AddProposal struct {
+	ClientInitKey ClientInitKey
+}
+
+type UpdateProposal struct {
+	// this is a HPKEPublicKey (for this structure purposes it is opaque)
+	LeafKey []byte `tls:"head=2"`
+}
+
+type RemoveProposal struct {
+	Removed uint32
+}
+
+type Proposal struct {
+	Type   ProposalType
+	Add    *AddProposal
+	Update *UpdateProposal
+	Remove *RemoveProposal
+}
+
+func (p Proposal) MarshalTLS() ([]byte, error) {
+	s := &Stream{}
```

Comment:
`Stream` should be upstreamed to the `mint/syntax` package eventually.

Instead of directly constructing, it would be better to make a `NewStream()` method.

Reply:
It might also be good to separate `ReadStream` and `WriteStream`.